### PR TITLE
refactor(mesh): replace spinner with skeleton

### PR DESCRIPTION
**Pull request by caesarWHLee**
# Notable Changes
- Refactor profile pages structure
- Update layout to contain desktop navigation
- Add skeletons
- Fix typescript version and adjust eslint config

## Refactor profile pages structure
- one layout for `profile/member` and `profile/publisher` -> each page contain its own layout to separate CSR (`profile/member`) and SSR pages (`profile/publisher`)
- put `profile/member/[customId]/follower` and `profile/member/[customId]/following` inside a routing group `(follow)` to share the same layout

## Update layout to contain desktop navigation
- rename prop `navigation` to `mobileNavigation`
- add prop `non-mobileNavigation` to config 

NonMobileNavigation 
截圖 2024-10-01 中午12 31 30

## Skeletons
Figma: https://www.figma.com/design/wqODc69zqgbu7TZFtQAmnN/2024_READr-Mesh_Web?node-id=9942-136692&t=cPabFyz0V0xz7gWQ-4
Add skeleton (Loading.tsx) for pages depending on it's CSR or SSR

### SSR 
Put `<Loading/>` inside `<LayoutTemplate/>` `suspenseFallback` prop:

Pages:
- homepage
- story
- profile/member/following
- profile/member/follower
- profile/publisher
- profile/publisher/follower
- point/sponsorship
- point/subscribe-stories

Testing:
Use this helper function to simulate slow API response form server side.
```
export function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

// in side page.tsx
await delay(3000)
```

### CSR 
Render `<Loading/>` when page data is loading

Pages:
- social
- profile/member
- point

### Hybrid
Use both SSR and CSR to show `<Loading/>`

Pages
- media

### Fix typescript version and adjust eslint config

**Issue:** 
```
➜  mesh git:(mesh-skeleton) ✗ yarn lint
yarn run v1.22.22
$ next lint
Failed to load config "next/core-web-vitals" to extend from.
Referenced from: /Users/caesar/Desktop/projects/Sachiel/.eslintrc.js
error Command failed with exit code 1.
```

**Observation:**
1. Issue happened after [PR839](https://github.com/readr-media/Sachiel/pull/839) merged.
2. Two TypeScript version installed from mesh package.json
```
typescript@^5:
  version "5.4.5"
  resolved "https://registry.yarnpkg.com/typescript/-/typescript-5.4.5.tgz#42ccef2c571fdbd0f6718b1d1f5e6e5ef006f611"
  integrity sha512-vcI4UpRgg81oIRUFwR0WSIHKt11nJ7SAVlYNIu+QpqeyXP+gpQJy/Z4+F0aGxSE4MqwjyXvW/TzgkLAx2AGHwQ==

typescript@^5.1:
  version "5.5.4"
  resolved "https://registry.yarnpkg.com/typescript/-/typescript-5.5.4.tgz#d9852d6c82bad2d2eda4fd74a5762a8f5909e9ba"
  integrity sha512-Mtq29sKDAEYP7aljRgtPOpTvOfbwRWlS6dPRzwjdE+C0R4brX/GUyhHSecbHMFLNBLcJIPt9nl9yG5TZ1weH+Q==
```
3. Move `next/core-web-vitals` extension from root `eslint.rc` to `mesh/eslint.rc` fix this error.
4. Current typescript (5.6.2) version is beyond `@typescript-eslint/typescript-estree`'s support
```
➜  mesh git:(mesh-skeleton) ✗ yarn lint
yarn run v1.22.22
$ next lint
=============

WARNING: You are currently running a version of TypeScript which is not officially supported by @typescript-eslint/typescript-estree.

You may find that it works just fine, or you may not.

SUPPORTED TYPESCRIPT VERSIONS: >=4.7.4 <5.5.0

YOUR TYPESCRIPT VERSION: 5.6.2

Please only submit bug reports when using the officially supported version.
```

**Solution:** 
Move `next/core-web-vitals` to packages and fix the typescript version below 5..5.0

### CODE REVIEW

**Comment by caesarWHLee** [comment]
改動範圍涉及各頁面，請 @nihclil @LinHeMa @tyc67 各字看一下各自頁面的 skeleton 和 `LayoutTemplate` 的改動，請 @erase2004 幫看一下 `next/core-web-vitals` 相關的 config 調整。

File: packages/mesh/app/point/layout.tsx [hunk 1]
```diff
@@ -41,7 +47,7 @@ export default function PointLayout({
           restrictMainWidth: true,
           footer: 'hidden sm:block',
         }}
-        navigation={{
+        mobileNavigation={{
           leftButtons: [
```
Comment (tyc67) [inline packages/mesh/app/point/layout.tsx]: 可以換 `<GoBackButton/>` ?

**Comment by nihclil** [comment]
LGTM 👍👍👍

**Review by nihclil — APPROVED** [review]

File: packages/mesh/tailwind.config.js [hunk 2]
```diff
@@ -20,12 +20,18 @@ module.exports = {
       width: {
         maxContent: '1440px',
         maxMain: '1120px',
+        maxDesktonNavigation: '680px',
```
Comment (tyc67) [inline packages/mesh/tailwind.config.js]: max"Desktop"Navigation ?

File: packages/mesh/components/layout-template/index.tsx [hunk 3]
```diff
@@ -21,34 +30,38 @@ type CustomStyle = {
...
-  {
-    type: 'stateless'
-  }
->
+  suspenseFallback?: React.ReactNode
+} & (
+  | {
+      type: 'default'
+      mobileNavigation?: MobileNavigationProps
+      nonMobileNavigation?: StatefulNavigationProps
```
Comment (tyc67) [inline packages/mesh/components/layout-template/index.tsx]: 如果用 `DefaultNavigationProps` 會不會增加可讀性 !?

Comment (caesarWHLee) [reply packages/mesh/components/layout-template/index.tsx]: 如果是覺得 nonMobile 不易懂，可以改名成 navigation，不過應該不會把 props 直接放在命名中，以和其他 props 命名對齊

File: packages/mesh/tailwind.config.js (hunk 2 above)
Comment (caesarWHLee) [reply packages/mesh/tailwind.config.js]: 修改在 [38e9ba0](https://github.com/readr-media/Sachiel/pull/855/commits/38e9ba0bb16115a6dbd7d5ec3525cc694e0a5a72)

File: packages/mesh/components/layout-template/index.tsx (hunk 3 above)
Comment (tyc67) [reply packages/mesh/components/layout-template/index.tsx]: 是因為看到 `StatefulHeader` -> `DefaultHeader`, 如果 `StatefulNavigationProps` 改為`DefaultNavigationProps` 不知道會不會增加可讀性

**Review by LinHeMa — APPROVED** [review]
LGTM

**Comment by caesarWHLee** [comment]
[9bb831a](https://github.com/readr-media/Sachiel/pull/855/commits/9bb831aaaf723740d4b2276c62ddee24b2b52877)

修改 `<LayoutTemplate/>` 中 `mobileNavigation` prop 的格式，比照 nonMobileNavigation，之後依照按鈕開發會逐步移除 `<MobileNavigationButton/>` 的使用。

File: packages/mesh/app/point/layout.tsx (hunk 1 above)
Comment (caesarWHLee) [reply packages/mesh/app/point/layout.tsx]: 感謝建議，mobileNavigation 架構調整，見 [comment](https://github.com/readr-media/Sachiel/pull/855#issuecomment-2390508768)

**Review by tyc67 — APPROVED** [review]

**Review by erase2004 — APPROVED** [review]
ESLint 相關調整經實測可以正確運作，LGTM 👏

File: packages/mesh/components/layout-template/index.tsx (hunk 3 above)
Comment (caesarWHLee) [reply packages/mesh/components/layout-template/index.tsx]: 懂了，這邊是我搞混命名了，應該要叫 `DefaultNavigationProps` 沒錯，感謝提醒，修改在 [8fec93c](https://github.com/readr-media/Sachiel/pull/855/commits/8fec93c65908fb5d5e9b84eb9f154e2292c3b13d)

**Pull request merged**